Trust final non-static fields

chrisvest opened the pull request:
- Un-final fields in classes that need to be `Serializable` – they don't need it because they're not used to publish state via a race.
- Remove `Serializable` from classes that don't need it.
- Add a jQAssistant check that fails the (`mvn install`) build when we have final instance fields in _our_ serializable classes – dependencies are unfortunately not checked.
- Add off-by-default JVM options to trust non-static final fields – it's a turbo button for when no dependency, plug-in or extension need to modify final instance fields via reflection or serialisation.

This modifies both kernel and core-edge code, so review from both teams would be nice.

chrisvest committed:
Make tests run with trust non-static final enabled

chrisvest committed:
Add JVM flags to wrapper.conf to trust non-static final fields

chrisvest committed:
Remove Serializable interface from classes that didn't appear to need it, and remove `final` modifier from fields in serializable classes

chrisvest committed:
Add jQAssistant rules for serializable cannot have final non-static fields

Also fix the dozen violations it found

chrisvest committed:
Make trusting non-static final fields turned off by default

The field team is expected to advise and guide customers in the use of this "turbo button".

chrisvest committed:
Fix a javadoc typo

apcj commented:
I'm happy with the changes in the core-edge module.

chrisvest commented:
Current status: This PR requires upgrades to our CI infrastructure, and won't turn green until that's done.

johan-neo commented:
Did we split this work in two? If we can not get an updated maven on the build get the change in then the tool to make sure we respect final later.

chrisvest committed:
Remove the jQAssistant reporting

It requires Maven 3.2 or newer, and our CI is not that up to date.
We may also want to look alternative tools to do this verification, since jQA cannot fully express what we want to disallow.

chrisvest committed:
Enable TrustNonStaticFinalFields by default

Reason being that extensions and embedded use will likely be a minority use-case in 3.0
